cu29_log_runtime/src/lib.rs

```diff
@@ -1,207 +1,248 @@
+
+        Self {
+            writer: BufWriter::new(writer),
+            bytes_written: 0,
+        }
+    }
+
+    pub fn bytes_written(&self) -> usize {
+        self.bytes_written
+    }
+
+    pub fn flush(&mut self) -> Result<(), EncodeError> {
+        self.writer.flush().map_err(|inner| EncodeError::Io {
+            inner,
+            index: self.bytes_written,
+        })
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use crate::CuLogEntry;
-    use bincode::config::standard;
-    use cu29_log::value::Value;
-
-    #[test]
-    fn test_encode_decode_structured_log() {
-        let log_entry = CuLogEntry {
-            time: 0.into(),
-            msg_index: 1,
-            paramname_indexes: vec![2, 3],
-            params: vec![Value::String("test".to_string())],
-        };
-        let encoded = bincode::encode_to_vec(&log_entry, standard()).unwrap();
-        println!("{:?}", encoded);
-        let decoded_tuple: (CuLogEntry, usize) =
-            bincode::decode_from_slice(&encoded, standard()).unwrap();
-        assert_eq!(log_entry, decoded_tuple.0);
+impl<W: Write> Writer for OwningIoWriter<W> {
+    #[inline(always)]
+    fn write(&mut self, bytes: &[u8]) -> Result<(), EncodeError> {
+        self.writer
+            .write_all(bytes)
+            .map_err(|inner| EncodeError::Io {
+                inner,
+                index: self.bytes_written,
+            })?;
+        self.bytes_written += bytes.len();
+        Ok(())
     }
 }
+
+
+/// This allows this crate to be used outside of Copper (ie. decoupling it from the unifiedlog.
+pub struct SimpleFileWriter {
+    path: PathBuf,
+    encoder: EncoderImpl<OwningIoWriter<File>, Configuration>,
+}
+
+impl SimpleFileWriter {
+    pub fn new(path: &PathBuf) -> CuResult<Self> {
+        let file = std::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(path)
+            .map_err(|e| format!("Failed to open file: {:?}", e))?;
+
+        let writer = OwningIoWriter::new(file);
+        let encoder = EncoderImpl::new(writer, bincode::config::standard());
+
+        Ok(SimpleFileWriter { path: path.clone(), encoder})
+    }
+}
+
+impl Debug for SimpleFileWriter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SimpleFileWriter for path {:?}", self.path)
+    }
+}
+
+impl WriteStream<CuLogEntry> for SimpleFileWriter {
+    #[inline(always)]
+    fn log(&mut self, obj: &CuLogEntry) -> CuResult<()> {
+        obj.encode(&mut self.encoder).map_err(|e| format!("Failed to write to file: {:?}", e))?;
+        Ok(())
+    }
+
+    fn flush(&mut self) -> CuResult<()> {
+        self.encoder.writer().flush().map_err(|e| format!("Failed to flush file: {:?}", e))?;
+        Ok(())
+    }
+}
+
+// #[cfg(test)]
+// mod tests {
+//     use crate::CuLogEntry;
+//     use bincode::config::standard;
+//     use cu29_log::value::Value;
+//
+//     #[test]
+//     fn test_encode_decode_structured_log() {
```

Comment:
Is this test meant to be revived or should it still be disabled?

Reply:
oh yes so sorry, I was postponing that because of the smallvecs... fixed!